File: go/roster.go
package wh40kdc

// Whole-army roster legality: the per-unit loadout check (loadout.go) layered
// with the nine army-construction dimensions. Go mirror of
// tools/src/data/roster-resolve.ts (validateRosterCore / NormRoster).

// normUnit is one unit in the normalised roster the core checker consumes.
// Empty-string enhancementID / leaderBodyguardID means "null".
type normUnit struct {
	unitID            string
	modelCount        int
	isWarlord         bool
	enhancementID     string
	leaderBodyguardID string
	keywordOverrides  []string
	counts            map[string]int
}

// normRoster is the compact roster spec shared by the runner op. factionID and
// battleSize are "" for null; forceDisposition is nil for null (distinct from
// the empty string, which is a picked-but-blank disposition).
type normRoster struct {
	factionID        string
	battleSize       string
	forceDisposition *string
	detachmentIDs    []string
	units            []normUnit
}

// unitLoadoutResult is the per-unit loadout verdict (the building block layered
// under the army checks). Carries the resolved unit's index and its violations.
type unitLoadoutResult struct {
	unitIndex  int
	violations []map[string]string
}

// rosterViolation is one army-level legality violation. unitIndex is -1 for an
// army-wide violation; severity is "error" or "warn".
type rosterViolation struct {
	code      string
	id        string
	unitIndex int
	severity  string
}

// validateRosterCore runs the per-unit loadout check on every resolved unit,
// then the nine army-construction dimensions. Returns the per-unit loadout
// results (resolved units, in source order) and the army violations (unsorted;
// the runner op assembles + sorts the final lines). Mirror of the TS reference.
func validateRosterCore(spec normRoster, ds *Dataset) ([]unitLoadoutResult, []rosterViolation) {
	var army []rosterViolation
	push := func(severity, code, id string, unitIndex int) {
		army = append(army, rosterViolation{code: code, id: id, unitIndex: unitIndex, severity: severity})
	}
	errV := func(code, id string, unitIndex int) { push("error", code, id, unitIndex) }

	resolveUnit := func(unitID string) *UnitView {
		if unitID == "" {
			return nil
		}
		if spec.factionID != "" {
			if uv, ok := ds.Units.GetInFaction(unitID, spec.factionID); ok {
				return uv
			}
		}
		if uv, ok := ds.Units.GetAny(unitID); ok {
			return uv
		}
		return nil
	}
	// The army faction's keywords ([Imperium, Adeptus Astartes, Blood Angels]
	// for a chapter): every unit in the faction's pool owns them — the
	// <CHAPTER>-style keyword that chapter-shared datasheet records can't
	// carry. Granted with the same subset rule that scopes a chapter's unit
	// pool, so allied units never gain them.
	var armyKeywords []string
	if spec.factionID != "" {
		if fac, ok := ds.Factions.Get(spec.factionID); ok {
			armyKeywords = getStrList(fac.Raw, "keywords")
		}
	}
	armyKeywordSet := map[string]struct{}{}
	for _, k := range armyKeywords {
		armyKeywordSet[k] = struct{}{}
	}
	keywordSet := func(view *UnitView, overrides []string) map[string]struct{} {
		s := map[string]struct{}{}
		for _, k := range getStrList(view.Raw, "keywords") {
			s[k] = struct{}{}
		}
		for _, k := range getStrList(view.Raw, "faction_keywords") {
			s[k] = struct{}{}
		}
		s[getStr(view.Raw, "name")] = struct{}{}
		if len(armyKeywordSet) > 0 {
			inPool := true
			for _, k := range getStrList(view.Raw, "faction_keywords") {
				if _, ok := armyKeywordSet[k]; !ok {
					inPool = false
					break
				}
			}
			if inPool {
				for _, k := range armyKeywords {
					s[k] = struct{}{}
				}
			}
		}
		for _, grantAny := range getList(view.Raw, "conditional_keywords") {
			grant, ok := asMap(grantAny)
			if !ok {
				continue
			}
			if requiredDetachmentID := getStr(grant, "required_detachment_id"); requiredDetachmentID != "" &&
				!contains(spec.detachmentIDs, requiredDetachmentID) {
				continue
			}
			if requiredFactionKeyword := getStr(grant, "required_faction_keyword"); requiredFactionKeyword != "" {
				if _, ok := armyKeywordSet[requiredFactionKeyword]; !ok {
					continue
				}
			}
			s[getStr(grant, "keyword")] = struct{}{}
		}
		for _, keyword := range overrides {
			s[keyword] = struct{}{}
		}
		return s
	}
	isCharacter := func(view *UnitView) bool {
		r := getStr(view.Raw, "role")
		return r == "character" || r == "epic-hero" || contains(getStrList(view.Raw, "keywords"), "Character")
	}

	views := make([]*UnitView, len(spec.units))
	for i, u := range spec.units {
		views[i] = resolveUnit(u.unitID)
	}

	// --- Per-unit loadout (reuse the tier/bounds checker). --------------------
	var units []unitLoadoutResult
	for idx, su := range spec.units {
		view := views[idx]
		if view == nil {
			continue
		}
		models, tiers := ds.unitCompositionOf(view.Raw)
		units = append(units, unitLoadoutResult{
			unitIndex:  idx,
			violations: checkUnitLegality(view.Raw, su.modelCount, ds.wargearOptionsOf(view.Raw), su.counts, models, tiers),
		})
	}

	// Resolved detachments (drop ids absent from the dataset).
	var detachments []map[string]any
	// Shared detachment ids (Codex chapters) resolve within the roster's
	// faction; fall back first-wins when the spec names no faction.
	for _, id := range spec.detachmentIDs {
		if detachment := lookupDetachment(ds, id, spec.factionID); detachment != nil {
			detachments = append(detachments, detachment)
		}
	}

	// --- Enhancements: per-unit eligibility + army-wide uniqueness. -----------
	enhUses := map[string]int{}
	for idx, su := range spec.units {
		if su.enhancementID == "" {
			continue
		}
		enhUses[su.enhancementID]++
		eAny, eok := ds.Enhancements.Get(su.enhancementID)
		view := views[idx]
		if !eok || view == nil {
			continue
		}
		enh := eAny.(map[string]any)
		enhID := getStr(enh, "id")
		if !contains(spec.detachmentIDs, getStr(enh, "detachment_id")) {
			errV("enhancement-wrong-detachment", enhID, idx)
		}
		if !isCharacter(view) && !contains(su.keywordOverrides, "Character") && enh["upgrade_tag"] != true {
			errV("enhancement-on-non-character", enhID, idx)
		}
		kws := keywordSet(view, su.keywordOverrides)
		eligible := true
		if rawGroups, present := enh["keyword_restriction_groups"]; present && rawGroups != nil {
			eligible = false
			groups, _ := asList(rawGroups)
			for _, rawGroup := range groups {
				group, _ := asList(rawGroup)
				groupEligible := true
				for _, rawKeyword := range group {
					keyword, ok := rawKeyword.(string)
					if !ok {
						groupEligible = false
						break
					}
					if _, has := kws[keyword]; !has {
						groupEligible = false
						break
					}
				}
				if groupEligible {
					eligible = true
					break
				}
			}
		} else {
			for _, keyword := range getStrList(enh, "keyword_restrictions") {
				if _, has := kws[keyword]; !has {
					eligible = false
					break
				}
			}
		}
		if !eligible {
			errV("enhancement-keyword-mismatch", enhID, idx)
		}
		for _, k := range getStrList(enh, "exclusion_keywords") {
			if _, has := kws[k]; has {
				errV("enhancement-excluded-keyword", enhID, idx)
				break
			}
		}
	}
	for enhID, uses := range enhUses {
		maxTargets := 1
		if eAny, ok := ds.Enhancements.Get(enhID); ok {
			if mt, has := eAny.(map[string]any)["max_targets"]; has && mt != nil {
				maxTargets = asInt(mt)
			}
		}
		if uses > maxTargets {
			errV("enhancement-over-max-targets", enhID, -1)
		}
	}

	// --- Leader attachment. ----------------------------------------------------
	for idx, su := range spec.units {
		view := views[idx]
		if view == nil {
			continue
		}
		if su.leaderBodyguardID != "" {
			eligible := bodyguardEligibleIDs(ds, view.ID())
			if su.enhancementID != "" {
				if enhancementAny, ok := ds.Enhancements.Get(su.enhancementID); ok {
					for _, bodyguardID := range getStrList(enhancementAny.(map[string]any), "attachment_bodyguard_ids") {
						eligible[bodyguardID] = struct{}{}
					}
				}
			}
			if _, ok := eligible[su.leaderBodyguardID]; !ok {
				errV("leader-attachment-illegal", view.ID(), idx)
			}
		} else if getStr(view.Raw, "attachment_role") == "support" &&
			(isCharacter(view) || contains(su.keywordOverrides, "Character")) {
			errV("leader-must-attach", view.ID(), idx)
		}
	}

	// --- Points total (ordinal-aware) + enhancement costs. --------------------
	// Host-aware: a foreign unit with an allied_points entry for this army
	// (Agents' Imperium price, a chapter's reprice of a shared datasheet)
	// prices from that entry, not its native table.
	var rosterFaction map[string]any
	if spec.factionID != "" {
		if f, ok := ds.Factions.Get(spec.factionID); ok {
			rosterFaction = f.Raw
		}
	}
	ordinals := map[string]int{}
	total := 0
	for idx, su := range spec.units {
		view := views[idx]
		if view == nil {
			continue
		}
		ord := ordinals[su.unitID] + 1
		ordinals[su.unitID] = ord
		total += hostUnitPoints(view.Raw, su.modelCount, ord, rosterFaction)
		total += wargearPoints(view.Raw, su.counts)
		if su.enhancementID != "" {
			if eAny, ok := ds.Enhancements.Get(su.enhancementID); ok {
				total += asInt(eAny.(map[string]any)["cost"])
			}
		}
	}
	if limit, ok := pointsLimitForBattleSize(spec.battleSize); ok && total > limit {
		errV("points-over-limit", "roster", -1)
	}

	// --- Detachment-point budget. ---------------------------------------------
	cap, capOk := detachmentCapForBattleSize(spec.battleSize)
	dpUsed := 0
	for _, d := range detachments {
		dpUsed += asInt(d["detachment_points"])
	}
	if capOk && dpUsed > cap {
		errV("detachment-points-over", "roster", -1)
	}

	// --- Force disposition (advisory / warn). ---------------------------------
	// Any selected detachment may grant the pick; detachments whose data does
	// not record force_dispositions (key absent or null) are skipped, and when
	// none record them the check is inconclusive and stays silent.
	if spec.forceDisposition == nil {
		push("warn", "disposition-not-picked", "roster", -1)
	} else {
		recorded, granted := false, false
		for _, d := range detachments {
			if fd, ok := d["force_dispositions"]; ok && fd != nil {
				recorded = true
				if contains(toStrList(fd), *spec.forceDisposition) {
					granted = true
				}
			}
		}
		if recorded && !granted {
			push("warn", "disposition-invalid", *spec.forceDisposition, -1)
		}
	}

	// --- Detachment tag uniqueness (one per shared tag). ----------------------
	tagCounts := map[string]int{}
	for _, d := range detachments {
		for _, t := range getStrList(d, "tags") {
			tagCounts[t]++
		}
	}
	for tag, n := range tagCounts {
		if n > 1 {
			errV("detachment-tag-conflict", tag, -1)
		}
	}

	// --- Detachment restrictions (required/excluded army keywords, per unit). -
	for _, d := range detachments {
		r, ok := getMap(d, "restrictions")
		if !ok {
			continue
		}
		required := getStrList(r, "required_keywords")
		excluded := getStrList(r, "excluded_keywords")
		for idx := range spec.units {
			view := views[idx]
			if view == nil {
				continue
			}
			kws := keywordSet(view, spec.units[idx].keywordOverrides)
			for _, k := range required {
				if _, has := kws[k]; !has {
					errV("detachment-restriction-required", view.ID(), idx)
					break
				}
			}
			for _, k := range excluded {
				if _, has := kws[k]; has {
					errV("detachment-restriction-excluded", view.ID(), idx)
					break
				}
			}
		}
	}

	// --- Faction exclusions (a generic unit barred from this army's chapter). --
	// The shared Space Marine pool can't drop a generic datasheet for one chapter,
	// so a removed-without-replacement unit (e.g. Librarians for Black Templars)
	// carries excluded_faction_keywords; it is illegal when the army's faction
	// keywords intersect that list. Mirror of TS unit-excluded-from-faction.
	if spec.factionID != "" {
		factionKeywords := map[string]struct{}{}
		if fac, ok := ds.Factions.Get(spec.factionID); ok {
			for _, k := range getStrList(fac.Raw, "keywords") {
				factionKeywords[k] = struct{}{}
			}
		}
		if len(factionKeywords) > 0 {
			for idx := range spec.units {
				view := views[idx]
				if view == nil {
					continue
				}
				for _, k := range getStrList(view.Raw, "excluded_faction_keywords") {
					if _, has := factionKeywords[k]; has {
						errV("unit-excluded-from-faction", view.ID(), idx)
						break
					}
				}
			}
		}
	}

	// --- Warlord present (exactly one). ---------------------------------------
	warlords := 0
	for _, su := range spec.units {
		if su.isWarlord {
			warlords++
		}
	}
	if warlords == 0 {
		errV("no-warlord", "roster", -1)
	} else if warlords > 1 {
		errV("multiple-warlords", "roster", -1)
	}

	// --- Unit minimums (e.g. Houndpack: 3+ WAR DOG units). --------------------
	for _, d := range detachments {
		for _, umAny := range getList(d, "unit_minimums") {
			um, _ := asMap(umAny)
			keyword := getStr(um, "keyword")
			minN := asInt(um["min"])
			count := 0
			for idx, v := range views {
				if v == nil {
					continue
				}
				if _, has := keywordSet(v, spec.units[idx].keywordOverrides)[keyword]; has {
					count++
				}
			}
			if count < minN {
				errV("unit-minimum-unmet", keyword, -1)
			}
		}
	}

	return units, army
}

// bodyguardEligibleIDs is the set of body-unit ids the given leader can attach
// to — its leader-attachment `eligible_bodyguard_ids` that resolve to a known
// unit. Mirror of Dataset.bodyguardsAttachableFrom (membership only).
func bodyguardEligibleIDs(ds *Dataset, leaderUnitID string) map[string]struct{} {
	out := map[string]struct{}{}
	for _, laAny := range ds.LeaderAttachments {
		la, _ := asMap(laAny)
		if getStr(la, "leader_id") != leaderUnitID {
			continue
		}
		for _, bid := range getStrList(la, "eligible_bodyguard_ids") {
			// Faction-agnostic attachment data — GetAny.
			if _, ok := ds.Units.GetAny(bid); ok {
				out[bid] = struct{}{}
			}
		}
	}
	return out
}

File: python/src/wh40kdc/data/roster.py
"""Whole-army roster legality: the per-unit loadout check plus the nine
army-construction dimensions layered over it (enhancements, leader attachment,
points, detachment points, force disposition, detachment tags/restrictions,
warlord, unit minimums).

A roster is legal iff its ``army`` list has no ``error``-severity entries and
every ``units[].violations`` is empty (force-disposition advisories are
``warn`` — provisional 11e data).

Python mirror of the roster-legality half of ``tools/src/data/roster-resolve.ts``
(``validateRosterCore`` / ``checkRoster``).
"""

from __future__ import annotations

from typing import Any

from wh40kdc.data.battle_sizes import (
    detachment_cap_for_battle_size,
    points_limit_for_battle_size,
)
from wh40kdc.data.dataset import Dataset
from wh40kdc.data.loadout import check_unit_legality
from wh40kdc.data.pricing import host_unit_points, wargear_points

#: Army-construction violation codes (distinct from per-unit loadout codes).
ROSTER_VIOLATION_CODES = (
    "enhancement-wrong-detachment",
    "enhancement-on-non-character",
    "enhancement-keyword-mismatch",
    "enhancement-excluded-keyword",
    "enhancement-over-max-targets",
    "leader-attachment-illegal",
    "leader-must-attach",
    "points-over-limit",
    "detachment-points-over",
    "disposition-not-picked",
    "disposition-invalid",
    "detachment-tag-conflict",
    "detachment-restriction-required",
    "detachment-restriction-excluded",
    "unit-excluded-from-faction",
    "no-warlord",
    "multiple-warlords",
    "unit-minimum-unmet",
)


def _keyword_set(
    unit: dict[str, Any],
    army_keywords: list[str],
    army_keyword_set: set[str],
    detachment_ids: list[str],
    keyword_overrides: list[str] | None = None,
) -> set[str]:
    """Return the roster-contextual keyword set used by construction checks."""
    owned = set(unit.get("keywords") or []) | set(unit.get("faction_keywords") or [])
    owned.add(unit.get("name"))
    faction_keywords = unit.get("faction_keywords") or []
    if army_keyword_set and all(keyword in army_keyword_set for keyword in faction_keywords):
        owned.update(army_keywords)
    for grant in unit.get("conditional_keywords") or []:
        required_detachment_id = grant.get("required_detachment_id")
        if required_detachment_id and required_detachment_id not in detachment_ids:
            continue
        required_faction_keyword = grant.get("required_faction_keyword")
        if required_faction_keyword and required_faction_keyword not in army_keyword_set:
            continue
        owned.add(grant["keyword"])
    owned.update(keyword_overrides or [])
    return owned


def _is_character(unit: dict[str, Any]) -> bool:
    role = unit.get("role")
    return role in ("character", "epic-hero") or "Character" in (unit.get("keywords") or [])


def validate_roster_core(spec: dict[str, Any], dataset: Dataset) -> dict[str, Any]:
    """The shared roster-legality core. Runs the per-unit loadout check on every
    resolved unit, then the nine army-construction dimensions.

    ``spec`` keys: ``faction_id`` (str | None), ``battle_size`` (str | None),
    ``force_disposition`` (str | None), ``detachment_ids`` (list[str]), and
    ``units`` — a list of dicts with ``unit_id``, ``model_count``,
    ``is_warlord`` (bool), ``enhancement_id`` (str | None),
    ``leader_bodyguard_id`` (str | None), and ``counts`` (id → int). A
    unit-scoped violation's ``unitIndex`` indexes ``spec["units"]``.

    Returns ``{"units": [UnitLegality], "army": [RosterViolation]}``.
    """
    faction_id = spec.get("faction_id")
    spec_units: list[dict[str, Any]] = spec.get("units") or []
    detachment_ids: list[str] = spec.get("detachment_ids") or []

    army: list[dict[str, Any]] = []

    def push(
        severity: str,
        code: str,
        id_: str,
        message: str,
        unit_index: int | None = None,
    ) -> None:
        army.append(
            {
                "code": code,
                "id": id_,
                "message": message,
                "unitIndex": unit_index,
                "severity": severity,
            }
        )

    def err(code: str, id_: str, message: str, unit_index: int | None = None) -> None:
        push("error", code, id_, message, unit_index)

    def resolve_unit(unit_id: str) -> Any:
        if not unit_id:
            return None
        if faction_id:
            scoped = dataset.units.get_in_faction(unit_id, faction_id)
            if scoped is not None:
                return scoped
        return dataset.units.get_any(unit_id)

    # The army faction's keywords ([Imperium, Adeptus Astartes, Blood Angels]
    # for a chapter): every unit in the faction's pool owns them — the
    # <CHAPTER>-style keyword that chapter-shared datasheet records can't
    # carry. Granted with the same subset rule that scopes a chapter's unit
    # pool, so allied units never gain them.
    army_keywords: list[str] = []
    if faction_id:
        army_faction = dataset.factions.get(faction_id)
        if army_faction is not None:
            army_keywords = list(army_faction.raw.get("keywords") or [])
    army_keyword_set = set(army_keywords)

    views = [resolve_unit(u.get("unit_id") or "") for u in spec_units]

    # --- Per-unit loadout (reuse the tier/bounds checker). --------------------
    units: list[dict[str, Any]] = []
    for idx, su in enumerate(spec_units):
        view = views[idx]
        if view is None:
            continue
        options = dataset.wargear_options_of(view.raw)
        composition = dataset.unit_composition_of(view.raw)
        units.append(
            {
                "unitId": view.id,
                "unitIndex": idx,
                "modelCount": su.get("model_count") or 0,
                "violations": check_unit_legality(
                    view.raw,
                    su.get("model_count") or 0,
                    options,
                    su.get("counts") or {},
                    (composition or {}).get("models"),
                    (composition or {}).get("tiers"),
                ),
            }
        )

    # Shared detachment ids (Codex chapters) resolve within the roster's
    # faction; fall back first-wins when the spec names no faction.
    detachments = [
        d
        for d in (
            (dataset.detachments.get_in_faction(id_, faction_id) if faction_id else None)
            or dataset.detachments.get_any(id_)
            for id_ in detachment_ids
        )
        if d is not None
    ]
    # --- Enhancements: per-unit eligibility + army-wide uniqueness. -----------
    enh_uses: dict[str, int] = {}
    for idx, su in enumerate(spec_units):
        enhancement_id = su.get("enhancement_id")
        if not enhancement_id:
            continue
        enh_uses[enhancement_id] = enh_uses.get(enhancement_id, 0) + 1
        enh = dataset.enhancements.get(enhancement_id)
        view = views[idx]
        if enh is None or view is None:
            continue
        if enh.get("detachment_id") not in detachment_ids:
            err(
                "enhancement-wrong-detachment",
                enh["id"],
                f"{enh['id']} is not from a detachment in this roster",
                idx,
            )
        overrides = su.get("keyword_overrides") or []
        if (
            not _is_character(view.raw)
            and "Character" not in overrides
            and enh.get("upgrade_tag") is not True
        ):
            err(
                "enhancement-on-non-character",
                enh["id"],
                f"{enh['id']} can only be taken by a Character",
                idx,
            )
        kws = _keyword_set(
            view.raw,
            army_keywords,
            army_keyword_set,
            detachment_ids,
            overrides,
        )
        groups = enh.get("keyword_restriction_groups")
        eligible = (
            any(all(keyword in kws for keyword in group) for group in groups)
            if groups is not None
            else all(keyword in kws for keyword in (enh.get("keyword_restrictions") or []))
        )
        if not eligible:
            err(
                "enhancement-keyword-mismatch",
                enh["id"],
                f"{view.id} lacks an eligible keyword group for {enh['id']}",
                idx,
            )
        if any(k in kws for k in (enh.get("exclusion_keywords") or [])):
            err(
                "enhancement-excluded-keyword",
                enh["id"],
                f"{view.id} carries a keyword excluded by {enh['id']}",
                idx,
            )
    for enh_id, uses in enh_uses.items():
        max_targets = (dataset.enhancements.get(enh_id) or {}).get("max_targets")
        if max_targets is None:
            max_targets = 1
        if uses > max_targets:
            err(
                "enhancement-over-max-targets",
                enh_id,
                f"{enh_id} taken {uses} times, max {max_targets}",
            )

    # --- Leader attachment. ----------------------------------------------------
    for idx, su in enumerate(spec_units):
        view = views[idx]
        if view is None:
            continue
        leader_bodyguard_id = su.get("leader_bodyguard_id")
        if leader_bodyguard_id:
            eligible_bodyguards = {v.id for v in dataset.bodyguards_attachable_from(view.id)}
            enhancement = (
                dataset.enhancements.get(su["enhancement_id"]) if su.get("enhancement_id") else None
            )
            if enhancement is not None:
                eligible_bodyguards.update(enhancement.get("attachment_bodyguard_ids") or [])
            if leader_bodyguard_id not in eligible_bodyguards:
                err(
                    "leader-attachment-illegal",
                    view.id,
                    f"{view.id} cannot attach to {leader_bodyguard_id}",
                    idx,
                )
        elif view.raw.get("attachment_role") == "support" and (
            _is_character(view.raw) or "Character" in (su.get("keyword_overrides") or [])
        ):
            err(
                "leader-must-attach",
                view.id,
                f"{view.id} is a Support character and must attach to a unit",
                idx,
            )

    # --- Points total (ordinal-aware) + enhancement costs. --------------------
    # Host-aware: a foreign unit with an ``allied_points`` entry for this army
    # (Agents' Imperium price, a chapter's reprice of a shared datasheet)
    # prices from that entry, not its native table.
    roster_faction_view = dataset.factions.get(faction_id) if faction_id else None
    roster_faction = roster_faction_view.raw if roster_faction_view else None
    ordinals: dict[str, int] = {}
    total = 0
    for idx, su in enumerate(spec_units):
        view = views[idx]
        if view is None:
            continue
        unit_id = su.get("unit_id") or ""
        ordinal = ordinals.get(unit_id, 0) + 1
        ordinals[unit_id] = ordinal
        total += host_unit_points(view.raw, su.get("model_count") or 0, ordinal, roster_faction)
        total += wargear_points(view.raw, su.get("counts") or {})
        enhancement_id = su.get("enhancement_id")
        if enhancement_id:
            total += (dataset.enhancements.get(enhancement_id) or {}).get("cost") or 0
    limit = points_limit_for_battle_size(spec.get("battle_size"))
    if limit is not None and total > limit:
        err("points-over-limit", "roster", f"army totals {total}, over the {limit} limit")

    # --- Detachment-point budget. ---------------------------------------------
    cap = detachment_cap_for_battle_size(spec.get("battle_size"))
    dp_used = sum((d.get("detachment_points") or 0) for d in detachments)
    if cap is not None and dp_used > cap:
        err(
            "detachment-points-over",
            "roster",
            f"detachments cost {dp_used} DP, over the {cap} budget",
        )

    # --- Force disposition (advisory / warn). ---------------------------------
    # Any selected detachment may grant the pick; detachments whose data does
    # not record force_dispositions are skipped, and when none record them the
    # check is inconclusive and stays silent.
    force_disposition = spec.get("force_disposition")
    if force_disposition is None:
        push("warn", "disposition-not-picked", "roster", "no Force Disposition selected")
    else:
        recorded = [d for d in detachments if d.get("force_dispositions") is not None]
        if recorded and not any(force_disposition in d["force_dispositions"] for d in recorded):
            push(
                "warn",
                "disposition-invalid",
                force_disposition,
                f"{force_disposition} is not offered by any selected detachment",
            )

    # --- Detachment tag uniqueness (one per shared tag). ----------------------
    tag_counts: dict[str, int] = {}
    for d in detachments:
        for t in d.get("tags") or []:
            tag_counts[t] = tag_counts.get(t, 0) + 1
    for tag, n in tag_counts.items():
        if n > 1:
            err("detachment-tag-conflict", tag, f"{n} detachments share the '{tag}' tag")

    # --- Detachment restrictions (required/excluded army keywords, per unit). -
    for d in detachments:
        restrictions = d.get("restrictions")
        if not restrictions:
            continue
        for idx, su in enumerate(spec_units):
            view = views[idx]
            if view is None:
                continue
            kws = _keyword_set(
                view.raw,
                army_keywords,
                army_keyword_set,
                detachment_ids,
                su.get("keyword_overrides") or [],
            )
            if any(k not in kws for k in (restrictions.get("required_keywords") or [])):
                err(
                    "detachment-restriction-required",
                    view.id,
                    f"{view.id} lacks a keyword required by {d['id']}",
                    idx,
                )
            if any(k in kws for k in (restrictions.get("excluded_keywords") or [])):
                err(
                    "detachment-restriction-excluded",
                    view.id,
                    f"{view.id} carries a keyword excluded by {d['id']}",
                    idx,
                )

    # --- Faction exclusions (a generic unit barred from this army's chapter). --
    # The shared Space Marine pool can't drop a generic datasheet for one chapter,
    # so a removed-without-replacement unit (e.g. Librarians for Black Templars)
    # carries ``excluded_faction_keywords``; it is illegal when the army's faction
    # keywords intersect that list. Mirror of TS ``unit-excluded-from-faction``.
    faction_view = dataset.factions.get(faction_id) if faction_id else None
    faction_keywords = set((faction_view.raw.get("keywords") if faction_view else None) or [])
    if faction_keywords:
        for idx, _su in enumerate(spec_units):
            view = views[idx]
            if view is None:
                continue
            barred = [
                k
                for k in (view.raw.get("excluded_faction_keywords") or [])
                if k in faction_keywords
            ]
            if barred:
                err(
                    "unit-excluded-from-faction",
                    view.id,
                    f"{view.id} cannot be taken by {faction_id} (barred by {', '.join(barred)})",
                    idx,
                )

    # --- Warlord present (exactly one). ---------------------------------------
    warlords = sum(1 for su in spec_units if su.get("is_warlord") is True)
    if warlords == 0:
        err("no-warlord", "roster", "army has no warlord")
    elif warlords > 1:
        err("multiple-warlords", "roster", f"army has {warlords} warlords")

    # --- Unit minimums (e.g. Houndpack: 3+ WAR DOG units). --------------------
    for d in detachments:
        for um in d.get("unit_minimums") or []:
            count = sum(
                1
                for idx, view in enumerate(views)
                if view is not None
                and um["keyword"]
                in _keyword_set(
                    view.raw,
                    army_keywords,
                    army_keyword_set,
                    detachment_ids,
                    spec_units[idx].get("keyword_overrides") or [],
                )
            )
            if count < um["min"]:
                err(
                    "unit-minimum-unmet",
                    um["keyword"],
                    f"{d['id']} requires {um['min']}+ {um['keyword']} units, found {count}",
                )

    army.sort(key=lambda v: (v["code"], v["id"]))
    return {"units": units, "army": army}


def check_roster(roster: dict[str, Any], dataset: Dataset) -> dict[str, Any]:
    """Whole-army legality for a resolved ``Roster`` dict. Lowers it to the
    normalised spec :func:`validate_roster_core` consumes (mirror of TS
    ``checkRoster``)."""
    units = []
    for u in roster.get("units") or []:
        counts: dict[str, int] = {}
        for w in u.get("wargear") or []:
            wid = w.get("ref", {}).get("id")
            if wid is None:
                continue
            counts[wid] = counts.get(wid, 0) + w.get("count", 0)
        leader_attachment = u.get("leader_attachment")
        leader_bodyguard_id = (
            leader_attachment.get("bodyguard_ref", {}).get("id") if leader_attachment else None
        )
        enhancement = u.get("enhancement")
        units.append(
            {
                "unit_id": (u.get("ref") or {}).get("id") or "",
                "model_count": u.get("model_count") or 0,
                "is_warlord": u.get("is_warlord") is True,
                "enhancement_id": enhancement.get("id") if enhancement else None,
                "leader_bodyguard_id": leader_bodyguard_id,
                "keyword_overrides": u.get("keyword_overrides") or [],
                "counts": counts,
            }
        )
    spec = {
        "faction_id": roster.get("faction_id"),
        "battle_size": roster.get("battle_size"),
        "force_disposition": roster.get("force_disposition"),
        "detachment_ids": [
            d["ref"]["id"]
            for d in (roster.get("detachments") or [])
            if (d.get("ref") or {}).get("id") is not None
        ],
        "units": units,
    }
    return validate_roster_core(spec, dataset)
